aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -18,6 +18,40 @@ static OffsetCalculator<N> index_make_offset_calculator(const TensorIterator& it
   return OffsetCalculator<N>(iter.ndim(), iter.shape().data(), strides.data());
 }
 
+__device__ char* device_strcpy(char* dst, const char* src, int n){
+  int i = 0;
+  while (i+1 < n && src[i]) {
+    dst[i] = src[i];
+    ++i;
+  }
+  if (i < n) {
+    dst[i] = '\0';
+  }
+  return dst;
+}
+
+__device__ bool _graceful_assert(c10::cuda::CUDAAssert* assert_state, bool condition, const char* message, uint32_t line, const char* file) {
+  if (!condition) { 
+    if (atomicCAS(&assert_state->error, 0, 1) == 0) {
+
+        // copy message
```

Comment:
It feels a little goofy that we literally copy the const char* pointer into assert state, when in principle it's just a string literal and the pointer should be good enough. I wonder if there's a way to share a constant between CUDA and CPU. Maybe we should just have a fixed set of error codes...

Reply:
It is true, the impl can be simplified in a couple of ways, e.g. by using only a fixed set of well-known error states. I am open to change the code in any direction. I wanted to keep generic / comparable to a CPU assert at first - but that might not be necessary.

I am also not really happy with copying strings in device code. I have not yet tested if it is possible to pass on pointers of string-constants defined in a cuda kernel. I will check this later today.